std/os/linux.zig

```diff
@@ -1213,10 +1224,50 @@ pub fn getsockopt(fd: i32, level: u32, optname: u32, noalias optval: [*]u8, noal
     return syscall5(SYS_getsockopt, @bitCast(usize, isize(fd)), level, optname, @ptrToInt(optval), @ptrToInt(optlen));
 }
 
-pub fn sendmsg(fd: i32, msg: *const msghdr, flags: u32) usize {
+pub fn sendmsg(fd: i32, msg: *msghdr_const, flags: u32) usize {
     return syscall3(SYS_sendmsg, @bitCast(usize, isize(fd)), @ptrToInt(msg), flags);
 }
 
+pub fn sendmmsg(fd: i32, msgvec: [*]mmsghdr_const, vlen: u32, flags: u32) usize {
+    if (@typeInfo(usize).Int.bits > @typeInfo(@typeOf(mmsghdr(undefined).msg_len)).Int.bits) {
+        // workaround kernel brokenness:
+        // if adding up all iov_len overflows a i32 then split into multiple calls
+        // see https://www.openwall.com/lists/musl/2014/06/07/5
+        const kvlen = if (vlen > IOV_MAX) IOV_MAX else vlen; // matches kernel
+        var next_unsent: usize = 0;
+        for (msgvec[0..kvlen]) |*msg, i| {
+            var size: i32 = 0;
+            const msg_iovlen = @intCast(usize, msg.msg_hdr.msg_iovlen); // kernel side this is treated as unsigned
+            for (msg.msg_hdr.msg_iov[0..msg_iovlen]) |iov, j| {
+                if (iov.iov_len > std.math.maxInt(i32) or @addWithOverflow(i32, size, @intCast(i32, iov.iov_len), &size)) {
+                    // batch-send all messages up to the current message
+                    if (next_unsent < i) {
+                        const batch_size = i - next_unsent;
+                        const r = syscall4(SYS_sendmmsg, @bitCast(usize, isize(fd)), @ptrToInt(&msgvec[next_unsent]), batch_size, flags);
+                        if (getErrno(r) != 0) return r;
```

Comment:
You should return `i` here when `i > 0`, otherwise the caller won't know how many messages have been processed. Also applies to lines 1252 and 1263.

This particular spot is a bit troublesome though because you're iterating over the iovecs; it results in a partially sent message when `j > 0`.

I'd just let sendmsg() (single m again :-)) deal with sum(iovecs) > INT_MAX.

Reply:
> This particular spot is a bit troublesome though because you're iterating over the iovecs; it results in a partially sent message when `j > 0`.

This is intentional. Such a case will return on the following line.

Reply:
Sorry, I probably didn't explain that well.

Say you have 4 messages: small, big, small, big (where big == overflows.) The order of events is this, if I read the source correctly:

1. Flush first 'small'
2. Parcel first 'big'
3. Flush second 'small'
4. Parcel second 'big'

If (3) errors, the caller won't know (1) and (2) succeeded because it receives an error code, not the number of messages sent.

Reply:
Hrm, right. it can know by looking in `msg.msg_len` for each message; but yes I guess it should return `next_unsent` here instead of `r`.